op-batcher/channel_manager.go

```diff
@@ -35,90 +69,174 @@ func (s *channelManager) Clear() {
 	s.datas = s.datas[:0]
 }
 
-// func (s *channelManager) TxConfirmed(id txID, inclusionBlock eth.BlockID) {
-// 	// todo: implement
-// }
+func (s *channelManager) TxConfirmed(id txID, inclusionBlock eth.BlockID) {
+	if _, ok := s.pendingTransactions[id]; !ok {
+		s.log.Info("marked transaction as confirmed despite having no record of it")
+		// TODO: This can occur if we clear the channel while there are still pending transactions
+		// We need to keep track of stale transactions instead
+		return
+	}
+	delete(s.pendingTransactions, id)
+	s.confirmedTransactions[id] = inclusionBlock
+
+	// If this channel timed out, put the pending blocks back into the local saved blocks
+	// and then reset this state so it can try to build a new channel.
+	if s.pendingChannelIsTimedOut() {
+		s.log.Warn("Channel timed out", "id", s.pendingChannel)
+		s.blocks = append(s.pendingBlocks, s.blocks...)
+		s.clearPendingChannel()
+	}
+	// If we are done with this channel, record that.
+	if s.pendingChannelIsFullySubmitted() {
+		s.log.Info("Channel is fully submitted", "id", s.pendingChannel)
+		s.clearPendingChannel()
+	}
+}
 
-// TxData returns the next tx.data that should be submitted to L1.
-// It is very simple & currently ignores the l1Head provided (this will change).
-// It may buffer very large channels as well.
-func (s *channelManager) TxData(l1Head eth.L1BlockRef) ([]byte, txID, error) {
-	// Note: l1Head is not actually used in this function.
+// TODO: Create separate "pending" state
+func (s *channelManager) clearPendingChannel() {
+	s.channelPending = false
+	s.pendingChannel = derive.ChannelID{}
+	s.pendingBlocks = nil
+	s.pendingTransactions = make(map[txID][]byte)
+	s.confirmedTransactions = make(map[txID]eth.BlockID)
+}
 
-	// Return a pre-existing frame if we have it.
-	if len(s.datas) != 0 {
-		r := s.datas[0]
-		s.datas = s.datas[1:]
-		return r.data, r.id, nil
+func (s *channelManager) pendingChannelIsTimedOut() bool {
+	if !s.channelPending {
+		return false // no channel to be timed out
+	}
+	min := uint64(math.MaxUint64)
+	max := uint64(0)
```

Comment:
I wish there was a cheap way to determine this by looking at the chain, having non persistent statefulness isn't ideal

When there are no confirmed txs:
```
0 - math.MaxUint64 >= s.channelTimeout
```

This will prob always be false?

Reply:
yea, I should probably short circuit if there are no confirmed transactions. I don't like taking my chances on the wrapping behavior here.